File: src/search/binarySearch.cs
using System;

namespace BSearch
{
    public class BSearch
    {
        private int[] _numArr;
        private int _arrSize;

        public int BinarySearch(int target) // Non-recursive binary search
        {
            int left = 0, right = this._arrSize - 1;
            int idx = -1;
            
            while (left <= right)
            {
                int mid = (left + right) / 2;
                if (_numArr[mid] == target)
                {
                    idx = mid;
                    break;
                }
                else if (target < _numArr[mid])
                    right = mid - 1;
                else left = mid + 1;
            }
            return idx;
        }

        public BSearch()
        {
            string[] input = Console.ReadLine().Split(' ', StringSplitOptions.RemoveEmptyEntries);
            this._numArr = new int[input.Length];
            this._arrSize = input.Length;
            for (var i = 0; i < this._arrSize; i++)
                _numArr[i] = Convert.ToInt32(input[i]);
            Array.Sort(this._numArr);
        }
    }

    public class Program
    {
        static void Main()
        {
            BSearch a = new();
            Console.WriteLine(a.BinarySearch(Convert.ToInt32(Console.ReadLine())));
        }
    }
}
    /*
        BinarySearch method returns -1 if the target is not exist in the array.
        Otherwise, it returns the index of target in array.
    */

    /*
        Input:
        5 4 3 2 1
        123
        Output:
        -1
    */

File: src/search/binarySearch.py
def binary_search(arr, size, target):
    left = 0
    right = size - 1
    idx = -1
    
    while left <= right:
        mid = (left + right) // 2
        if target < arr[mid]:
            right = mid - 1
        elif arr[mid] < target:
            left = mid + 1
        else:
            idx = mid
            break
    return idx


def print_array(arr, size):
    for i in range(size):
        print(arr[i], end=' ')
    print()


arr = list(map(int, input().split(' ')))
size = len(arr)
print_array(arr, size)
arr.sort()
print_array(arr, size)

target = int(input("Target: "))
rt = binary_search(arr, size, target)

print("The target", end=' ')
if rt == -1:
    print(target, "is NOT exist.")
else:
    print(target, "is exist at index", rt, "in the array.")

'''
Input:
5 4 3 2 1
Target: 3

Output:
5 4 3 2 1
1 2 3 4 5
3 is exist at index 2 in the array.
'''